core/src/main/java/jenkins/util/AntClassLoader.java

```diff
@@ -772,6 +773,28 @@ private InputStream loadBaseResource(String name) {
         return parent == null ? super.getResourceAsStream(name) : parent.getResourceAsStream(name);
     }
 
+    private interface Convert<T> {
+        T convert(@Nonnull JarFile jarFile, @Nonnull JarEntry entry) throws IOException;
+    }
+
+    private <T> T storeAndConvert(JarFile jarFile, File file, String resourceName, Convert<T> convert) throws IOException {
+        if (jarFile == null) {
+            if (file.exists()) {
+                jarFile = new JarFile(file);
+                jarFiles.put(file, jarFile);
+            } else {
+                return null;
+            }
+            //to eliminate a race condition, retrieve the entry
```

Comment:
Does this comment really represent the logic? it was there for one code part, but not for another one. For me it looks like there was missing `TODO` here

Reply:
I _think_ that this code is how it tries to eliminate a race condition.

#### The simplest racing logic would be:
```
 if (jarFile == null) {
            if (file.exists()) {
                jarFile = new JarFile(file);
                jarFiles.put(file, jarFile);
            } else {
                return null;
            }
// at this point, you'd expect `jarFile == (JarFile) jarFiles.get(file)`, but it might not if you win the race.
}
```

The correct thing to do is to lock something. Otherwise, even this code probably isn't good enough.

### I think w/ this code and 3 threads we could still get a race:

#### Enter the if:
```
(1) T1  if (jarFile == null) {
(2) T2  if (jarFile == null) {
(3) T3  if (jarFile == null) {
```
#### Create a file:
```
(4) T1                 jarFile = new JarFile(file);
(5) T2                 jarFile = new JarFile(file);
(6) T3                jarFile = new JarFile(file);
```
#### Race time:
```
(7) T1                jarFiles.put(file, jarFile);
(8) T2                jarFiles.put(file, jarFile);
(9) T1           jarFile = (JarFile) jarFiles.get(file);
(10) T2           jarFile = (JarFile) jarFiles.get(file);
(11) T3                jarFiles.put(file, jarFile);
(12) T3           jarFile = (JarFile) jarFiles.get(file);
```
#### Check our winner's:
```
(13) T1        JarEntry entry = jarFile.getJarEntry(resourceName);
This is T2's jarFile from (5)/(8) -- retrieved at (9)
(14) T2        JarEntry entry = jarFile.getJarEntry(resourceName);
This is T2's jarFile from (5)/(8) -- stored at (8) and retrieved at (10)
(15) T3        JarEntry entry = jarFile.getJarEntry(resourceName);
This is T3's jarFile from (6) -- stored at (11) and retrieved at (12)
```

Reply:
I do *not* want to change the logic in this commit / PR. Changing logic should not really be done in refactoring.

It makes it *really* hard to review / look back at things retroactively / debug regressions / etc.